lib/ahoy_activity_utils.rb

```diff
@@ -1,83 +1,92 @@
-        puts "\tUser ID: #{user}"
+      # We group by collection first, so we can sort and sum the timestamps
+      events.group_by { |e| e.properties['collection_id'] }.each do |cid, event|
+        timestamps = event.map { |e| [e[:time], e[:name]] }
 
-        # We figure if you are logged in and doing stuff it counts as volunteer time
-        # This could be expanded or resctricted
-        events = Ahoy::Event
-            .where(user_id: user)
-            .where("time BETWEEN ? AND ?", day.beginning_of_day, day.end_of_day)
-            .select(:id, :user_id, :time, :name, :properties)
+        minutes = total_contiguous_seconds(timestamps) / 60
 
-        unless events.empty?
-            # We group by collection first, so we can sort and sum the timestamps
-            events
-                .group_by { |e| e.properties["collection_id"] }
-                .each do |cid, event|
-                    timestamps = event.map{|e| [e[:time], e[:name]] }
+        next unless minutes.positive?
 
-                    minutes = self.total_contiguous_seconds(timestamps) / 60
-                    
-                    unless minutes <= 0
-                        begin 
-                            activity = AhoyActivitySummary.create({
-                                date: day.beginning_of_day,
-                                user_id: user,
-                                collection_id: cid,
-                                activity: 'transcribe',
-                                minutes: minutes
-                            })
-                        rescue ActiveRecord::RecordNotUnique => e
-                            puts e
-                        else
-                            puts "\t\t- Collection: #{activity.collection_id} \t#{activity.minutes} minutes\n"
-                        end
-                    end
-                end
-            end
+        begin
+          activity = AhoyActivitySummary.create(
+            {
+              date: day.beginning_of_day,
+              user_id: user,
+              collection_id: cid,
+              activity: 'transcribe',
+              minutes: minutes
+            }
+          )
+        rescue ActiveRecord::RecordNotUnique => e
+          puts e
+        else
+          puts "\t\t- Collection: #{activity.collection_id} \t#{activity.minutes} minutes\n"
         end
+      end
     end
 
-    def self.total_contiguous_seconds(times_and_names, tolerance=90.minutes)
-        total_seconds = 0
-        from_time = nil
+    latest_ahoy_activity_summary = AhoyActivitySummary.order(date: :desc).first
+    keep_after_date = [AhoyActivitySummary::KEEP_AFTER.ago.to_date, latest_ahoy_activity_summary.date.yesterday].min
+
+    exclude_actions = AhoyActivitySummary::WEEKLY_TRIAL_COHORT_TARGET_ACTIONS +
+                      AhoyActivitySummary::WEEKLY_TRANSCRIBER_COHORT_TARGET_ACTIONS
+
+    Ahoy::Event.where('time < ?', keep_after_date).where.not(name: exclude_actions).destroy_all
```

Comment:
I tried running this locally in my development environment, and the process ran for an hour and got up to 15GB of memory before I killed it.

Can we switch these two lines to [call `delete_all` instead of `destroy_all`](https://stackoverflow.com/questions/6698207/delete-all-vs-destroy-all)?

Reply:
Done. Although, without proper callbacks we are producing orphaned records this way (I believe deeds will be affected). Are we sure that is okay?

Reply:
That's  a good point.  Ahoy events should be leaf nodes in the data model, and Ahoy Visits without any events should also be leaf nodes.  Let me take a look at the deed data model and think harder about this.

Reply:
Okay, `Deed` has a foreign key to `Ahoy::Visit`, which we use when showing which deeds were created during a particular session on the Admin->Users->Visits listing.  We never travel in the other direction in the application, so we shouldn't run into a nil `visit` when displaying a deed.

That said, it would be nice to clean these up. Perhaps we could get the list of `visit_ids` which we are planning to delete, call `delete_all` on them, then call `Deed.where(visit_id: visit_ids_to_delete).update_all(visit_id: nil)`?

Reply:
Can you please try again and see how is the performance with this one.

We do destroy_all in batches. Maybe that would help.

Reply:
Please git pull --rebase or reset to head if there are any conflicts